### ISSUE TITLE
 DefaultObjectPool<T>.Return() is not thread-safe

### ISSUE THREAD

**Issue by gmajian** [aspnet/Common#273]
DefaultObjectPool<T>.Get() is  thread-safe, but  DefaultObjectPool<T>.Return() is not thread-safe

```    public override void Return(T obj)
        {
            if (!_policy.Return(obj))
            {
                return;
            }

            for (var i = 0; i < _items.Length; i++)
            {
                if (_items[i] == null)  
                {
                    _items[i] = obj; // this may overwrite the non-null item
                    return;
                }
            }
        }
```

**Comment by Itfly** [comment]
When two threads within separate object call the Return method same time and go into the if sentence with same i, one of the object is not return and what's worse it causes memory leak.
For now, we have to use a lock before calling the Return method to fix this problem. Could you fix it in the library.

**Comment by oakio** [comment]
https://github.com/aspnet/Common/pull/282

**Comment by rynowak** [comment]
> When two threads within separate object call the Return method same time and go into the if sentence with same i, one of the object is not return and what's worse it causes memory leak.

Can you describe what you mean by memory leak? The design intent is that we allow one of the items be GC-ed in this rare case rather than forcing synchronization. That's not a leak.

If you're seeing this frequently or if it is genuinely a leak then that's a problem

**Comment by Itfly** [comment]
As gmajian say, the sentence "_items[i] = obj; " could be overwrite the non-null item. And worse that overwrited item can not be Disposed which cause memory leak.

**Comment by rynowak** [comment]
> Then that overwrited item can not be Disposed which cause memory leak.

The object pool **never** disposes items. If you are putting disposables in, then you can safely expect `Dispose` **never** to be called. If you are implementing `IDisposable` objects which hold native memory you need to implement a finalizer - or even better, use a `SafeHandle` http://blog.benoitblanchon.fr/safehandle/

Memory leaks involve an object becoming unreachable, but not eligible for collection. If you hit a race condition here the object that is dropped is eligible for collection. 

-----------------

For the record, I'm ok with us changing this behavior to be more deterministic so it's less surprising. 

If this really were a memory leak we would consider patching it.

**Comment by Itfly** [comment]
I implement the IPooledObjectPolicy interface and the Return method would dispose the obj if the actual size large than the pool capacity, here's code:
`public bool Return(TeeAnonymizer obj)
        {
            if (_counter > MaximumRetainedCapacity)
            {
                obj.Dispose();
                _counter--;
                return false;
            }

            return true;
        }`

The obj do not hold native memory but has some large objects(dictionaries, lists, files) in it. Yes it's a race condition, but the obj will not be dropped immediately and the memory increases rapidly because of creating new instances since the obj is not returned successfully.
And another problem for me is that it costs time (500ms or more) to create a new instance.

**Comment by rynowak** [comment]
This has been fixed by 07f1348
